File: 4/longestCommonPrefix.py
# leetCode question, ans by saketh
# Tue 17 Dec 12:38:14 AEDT 2019

def longestCommonPrefix(strs):
    out = strs[0] 
    counter = 0

    for i in range(1, len(strs)):
        newout = []
        #print(f'out = {out}, strs[i] = {strs[i]}')
        for j in range(min(len(strs[i]), len(out))):
            #print(f'j = {j}, strs[i][j] = {strs[i][j]}, out[j] = {out[j]}')
            if(out[j] != strs[i][j]):
                if ( j == 0 ):
                    return ""
                #else :
                #    newout = out[:j]
            else :
                newout.append(out[j])
        out = newout
    return ''.join(out)

                    

print(longestCommonPrefix(["flower","flow","flight", "fly"]))
print(longestCommonPrefix(["flower","flow","flight"]))
print(longestCommonPrefix(["dog","racecar","car"]))

File: tsPractice/longestCommonPrefix.ts
function longestCommonPrefix(strs: string[]): string {
    var com = '';
    var char = '';

    if (strs.length === 0) {
        return com;
    } else if (strs.length === 1) {
        return strs[0];
    } else {

        for (var j = 0; j < strs.length; ++j) {
            if (strs[j].length === 0) {
                return com;
            }

        };
    }
    // while (true) {
    for (var i = 0; i < strs[0].length; ++i) {
        char = strs[0].charAt(com.length);
        //console.log(char);
        for (var j = 0; j < strs.length; ++j) {
            if (char != strs[j].charAt(com.length)) {
                //console.log('here');
                return com;
            }

        };
        com += char;
        //console.log(com);
    }
    return com;
}

console.log('ans: ' + longestCommonPrefix(["flower", "flow", "floight"]));
console.log('ans: ' + longestCommonPrefix(["dog", "racecar", "car"]));
console.log('ans: ' + longestCommonPrefix(["", "", ""]));
console.log('ans: ' + longestCommonPrefix(['a']));
